tools/kill_if_no_output.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
+import fcntl
+import os
+import select
+import signal
+import subprocess
+import sys
+from typing import Any, List
+
+
+def make_async(fd: int) -> None:
+    fcntl.fcntl(fd, fcntl.F_SETFL, fcntl.fcntl(fd, fcntl.F_GETFL) | os.O_NONBLOCK)
+
+
+def kill_child(process, dormant_signal: int, kill_after: float, kill_signal: int) -> None:
+    os.kill(process.pid, dormant_signal)
+
+    try:
+        # process.wait handles timeouts and captures the exit status
+        process.wait(timeout=kill_after)
+    except subprocess.TimeoutExpired:
+        os.kill(process.pid, kill_signal)
+
+
+def read_until_exhaustion(fd: int) -> None:
+    try:
```

Comment:
Nitpick: Why duplicate this block outside of the loop?

Reply:
the loop is simpler this way. Do you have a rewrite suggestion?

Reply:
How about:
```python
while True:
    try:
        data = os.read(fd, 1024)
    except BlockingIOError:
        break
    sys.stdout.buffer.write(data)
    sys.stdout.flush()
```

Reply:
That was my first tried, it became an infinite loop. It looks like the call `os.read` will return empty once the file is closed, and a call to a closed file blocks. Maybe there are missing flags for the fcntl call

Reply:
Btw, I tested the python script with this shell script, if you want to have a try:

```sh
#!/usr/bin/bash

trap 'echo "Caught SIGUSR1"; sleep 5' SIGUSR1

>&2 echo bad
echo good

sleep 5

>&2 echo bad2
echo good2
```